rfcs/2020-03-06-1999-api-extensions-for-lua-transform.md

```diff
@@ -0,0 +1,204 @@
+
+### Proposed changes
+
+* Add `version` configuration option which would allow the users to chose between the new API described in this RFC (version 2) and the old one (version 1).
+* Hooks for initialization and shutdown called `init` and `shutdown`. They are defined as strings of Lua code in the `hooks` section of the configuration of the transform.
+* Timers which define pieces of code that are executed periodically. They are defined in array `timers`, each timer takes two configuration options: `interval` which is the interval for execution in seconds and `source` which is the code which is to be executed periodically.
+* Events are produced by the transform by calling function `emit` with the first argument being the event and the second option argument being the name of the lane where to emit the event. Outputting the events by storing them to the `event` global variable should not be supported, so its content would be ignored.
+* Support direct access to the nested fields (in both maps and arrays).
+* Add support for the timestamp type as a `userdata` object with the same visible fields as in the table returned by [`os.date`](https://www.lua.org/manual/5.3/manual.html#pdf-os.date). In addition, monkey-patch `os.date` function available inside Lua scripts to make it return the same kind of userdata instead of a table if it is called with `*t` or `!*t` as the argument. This is necessary to allow one-to-one correspondence between types in Vector and Lua.
+
+## Sales Pitch
+
+The proposal
+
+* gives users more power to create custom transforms;
+* supports both logs and metrics;
+* makes it possible to add complexity to the configuration of the transform gradually only when needed.
+
+## Drawbacks
+
+The only drawback is that supporting both dot notation and classical indexing makes it impossible to add escaping of dots in field names. For example, for incoming event structure like
+
+```json
+{
+  "field.first": {
+    "second": "value"
+  }
+}
+```
+
+accessing `event["field.first"]` would return `nil`.
+
+However, because of the specificity of the observability data, there seems to be no need to have both field names with dots and nested fields.
+
+## Outstanding Questions
+
+* Should timestamps be automatically inserted to created logs and metrics created as tables inside the transform is they are not present?
+* Are there better alternatives to the proposed solution for supporting of the timestamp type?
+* Could some users be surprised if the transform which doesn't call `emit` function doesn't output anything?
+* `null` might present in the events would be lost because in Lua setting a field to `nil` means deletion. Is it acceptable? If it is not, it is possible to introduce a new kind of `userdata` for representing `null` values.
```

Comment:
What's your take on this @binarylogic ?

Reply:
That's odd, but I think this is fine since it is a common Lua behavior.